### ISSUE TITLE
v2.0.x not working in production build with es2015

### ISSUE THREAD

**Issue by atlesj** [rintoj/ngx-virtual-scroller#323]
It works fine if targeting es5, and also if targeting es2015 and setting buildOptimizer = false in config. But using es2015 and buildOptimizer = true (as is default in production) doesn't work.

It compiles and renders, but you can't scroll to the virtual elements in the list, and when trying it gives a runtime error.

This can easily be reproduced in the demo project in this repo, by changing tsconfig target to es2015 and running ng serve --prod.

The problem might be that the build optimizer strips some vital parts away, but it seems to run fine with the rest of our project, so we have "solved" this by downgrading this component to 1.0.17, since that doesn't include an es2015 version in npm.

**Comment by speige** [comment]
Thanks for reporting this. I'm able to reproduce it.

It appears this function is being optimized out:
```
protected onScroll: () => void;
```

I'm not sure how to fix it yet. Still researching.

**Comment by speige** [comment]
On 2nd glance, there are bunch of lines of code which are optimized out. This causes this.onScroll to be undefined, hence your exception. However, once that's fixed there will be many more exceptions for similar reasons.

Example:
```
	public set scrollThrottlingTime(value: number) {
		this._scrollThrottlingTime = value;
		this.updateOnScrollFunction(); //optimized out
	}

	public set scrollDebounceTime(value: number) {
		this._scrollDebounceTime = value;
		this.updateOnScrollFunction(); //optimized out
	}

	public set checkResizeInterval(value: number) {
		if (this._checkResizeInterval === value) {
			return;
		}

		this._checkResizeInterval = value; //optimized out
		this.addScrollEventHandlers();
	}
```

You can see this by debugging the optimized demo & letting getting the minified source & running beautifier on it & comparing to the original source.

I'm not sure if it's Angular-CLI, or Terser (replacement for UglifyJS), or something else that's optimizing this out. But, looking at the node_modules\ngx-virtual-scroller folder I can see the line is not yet optimized out.

The big question is "why" is it optimized out? This is unfamiliar territory for me, so I'm not sure where to go next.

It feels like a bug in AngularCLI, but probably there's something I'm doing wrong. Maybe there's an extra flag I have to do add to prevent important methods from being removed?

I'll do some googling, but if I can't find anything we'll need to ask an Angular expert (or re-post this to the AngularCLI github issues) & see if we can get some clarification. 

If you're able to find anything that'll help in researching this, please let me know. Thanks!

**Comment by speige** [comment]
Ok, I found the flag that prevents this.

in package.json, I added:
"sideEffects': true

Essentially without this code it assumes all get/set properties are pure fields with no side-effects, which isn't true for ngx-virtual-scroller.

The issue should now be resolved. I've published version 2.0.7 to npm.

Please re-open if you continue to have issues.

**Comment by atlesj** [comment]
Wow, that was fast! Thanx
________________________________
